net/sched/p4tc/p4tc_table.c

```diff
@@ -603,26 +709,177 @@ tcf_table_find_byany(struct p4tc_pipeline *pipeline, struct nlattr *name_attr,
+		if (ret < 0) {
+			kfree(default_acts);
+			goto default_act_free;
+		} else if (ret > 1) {
+			NL_SET_ERR_MSG(extack, "Can only have one hit action");
+			ret = -EINVAL;
+			goto default_acts_free;
+		}
+	}
+
+	if (tb[P4TC_TABLE_DEFAULT_PERMISSIONS]) {
+		__u16 *permissions;
+
+		permissions = nla_data(tb[P4TC_TABLE_DEFAULT_PERMISSIONS]);
+		if (*permissions > (1 << 10) - 1) {
+			NL_SET_ERR_MSG(extack,
+				       "Permission may only have 10 bits turned on");
+			ret = -EINVAL;
+			goto default_acts_free;
+		}
+		if (!(*permissions & P4TC_PERMISSIONS_DATA_X)) {
+			NL_SET_ERR_MSG(extack,
+				       "Default action must have data path execute permissions");
+			ret = -EINVAL;
+			goto default_acts_free;
+		}
+		(*default_act)->permissions = *permissions;
+	} else {
+		(*default_act)->permissions = curr_permissions;
+	}
+
+	return 0;
+
+default_acts_free:
+	if (*default_act && (*default_act)->default_acts) {
+		tcf_action_destroy((*default_act)->default_acts,
+				   TCA_ACT_UNBIND);
+		kfree((*default_act)->default_acts);
+	}
+
+default_act_free:
+	kfree(*default_act);
+
+	return ret;
+}
+
+static int
+tcf_table_init_default_acts(struct net *net, struct nlattr **tb,
+			    struct p4tc_table *table,
+			    struct p4tc_table_default_action **default_hitact,
+			    struct p4tc_table_default_action **default_missact,
+			    struct netlink_ext_ack *extack)
+{
+	struct nlattr *tb_default[P4TC_TABLE_DEFAULT_MAX + 1];
+	__u16 permissions = P4TC_PERMISSIONS_CONTROL | P4TC_PERMISSIONS_DATA;
+	int ret;
+
+	*default_missact = NULL;
+	*default_hitact = NULL;
+
+	if (tb[P4TC_TABLE_DEFAULT_HIT]) {
+		struct p4tc_table_default_action *defact;
+
+		rcu_read_lock();
+		if (table) {
+			defact = rcu_dereference(table->tbl_default_hitact);
+
+			if (defact)
+				permissions = defact->permissions;
+		}
+		rcu_read_unlock();
+
+		ret = nla_parse_nested(tb_default, P4TC_TABLE_DEFAULT_MAX,
+				       tb[P4TC_TABLE_DEFAULT_HIT], NULL,
+				       extack);
+		if (ret < 0)
+			return ret;
+
+		ret = tcf_table_init_default_act(net, tb_default,
+						 default_hitact, permissions,
+						 extack);
+		if (ret < 0)
+			return ret;
+	}
+
+	if (tb[P4TC_TABLE_DEFAULT_MISS]) {
+		struct p4tc_table_default_action *defact;
+
+		rcu_read_lock();
+		if (table) {
+			defact = rcu_dereference(table->tbl_default_missact);
+
+			if (defact)
+				permissions = defact->permissions;
+		}
+		rcu_read_unlock();
+		ret = nla_parse_nested(tb_default, P4TC_TABLE_DEFAULT_MAX,
+				       tb[P4TC_TABLE_DEFAULT_MISS], NULL,
+				       extack);
+		if (ret < 0)
+			goto default_hitacts_free;
+
+		ret = tcf_table_init_default_act(net, tb_default,
+						 default_missact, permissions,
+						 extack);
+		if (ret < 0)
+			goto default_hitacts_free;
+	}
+
+	return 0;
+
+default_hitacts_free:
+	if (*default_hitact && (*default_hitact)->default_acts) {
+		tcf_action_destroy((*default_hitact)->default_acts,
+				   TCA_ACT_UNBIND);
+		kfree((*default_hitact)->default_acts);
+	}
```

Comment:
I see that you have this sort of code structure in some functions, can't we simplify the allocations/error handling?

Reply:
Tried to change the order and put this last to avoid doing the same thing in `tcf_table_init_default_action` (https://github.com/p4tc-dev/linux-p4tc-pub/pull/33/files#diff-feeeb2f6aa78dcb7c3d03d9ca086c03d2b3e2c38896b50419b9cc556725a4866R747-R778)